cmd/av/stack_sync.go

```diff
@@ -0,0 +1,280 @@
+
+If the --trunk flag is given, this command will synchronize changes from the
+latest commit to the repository base branch (e.g., main or master) into the
+stack. This is useful for rebasing a whole stack on the latest changes from the
+base branch.
+`),
+	RunE: func(cmd *cobra.Command, args []string) error {
+		repo, err := getRepo()
+		if err != nil {
+			return err
+		}
+
+		diff, err := repo.Diff(&git.DiffOpts{Quiet: true})
+		if err != nil {
+			return err
+		}
+		if !diff.Empty {
+			return errors.New("refusing to sync: there are unstaged changes in the working tree")
+		}
+		logrus.Debugf("%#+v", diff)
+
+		config := stackSyncFlags.stackSyncConfig
+
+		// check for --continue/--abort
+		if stackSyncFlags.Continue && stackSyncFlags.Abort {
+			return errors.New("cannot use --continue and --abort together")
+		}
+		existingState, err := readStackSyncState(repo)
+		if err != nil && !os.IsNotExist(err) {
+			return err
+		}
+		if stackSyncFlags.Abort {
+			if existingState.CurrentBranch == "" {
+				return errors.New("no sync in progress")
+			}
+			err := writeStackSyncState(repo, nil)
+			if err != nil {
+				return errors.Wrap(err, "failed to reset stack sync state")
+			}
+			_, _ = fmt.Fprintf(os.Stderr, "Aborted stack sync for branch %q\n", existingState.CurrentBranch)
+			return nil
+		} else if stackSyncFlags.Continue {
+			if existingState.CurrentBranch == "" {
+				return errors.New("no sync in progress")
+			}
+			config = existingState.Config
+			config.Continue = true
+		} else {
+			if existingState.CurrentBranch != "" {
+				return errors.New("a sync is already in progress: use --continue or --abort")
+			}
+		}
+
+		conflict, err := doStackSync(repo, config)
+		if err != nil {
+			return err
+		}
+		if conflict {
+			currentBranch, err := repo.CurrentBranchName()
+			if err != nil {
+				return err
+			}
+			if err := writeStackSyncState(repo, &stackSyncState{
+				CurrentBranch: currentBranch,
+				Config:        config,
+			}); err != nil {
+				return errors.Wrap(err, "failed to write stack sync state")
+			}
+			return errors.New("conflict detected: please resolve and then run `av stack sync --continue`")
+		}
+
+		if err := writeStackSyncState(repo, nil); err != nil {
+			return errors.Wrap(err, "failed to reset stack sync state")
+		}
+
+		_, _ = fmt.Fprintf(os.Stderr, "Stack sync complete\n")
+		return nil
+	},
+}
+
+// doStackSync performs the stack sync operation.
+// It returns a boolean indicating whether or not the sync was interrupted
+// (if true, the stack sync can be resumed with the --continue flag).
+func doStackSync(repo *git.Repo, config stackSyncConfig) (bool, error) {
+	root, err := stacks.GetCurrentRoot(repo)
+	if err != nil {
+		return false, err
+	}
+
+	currentBranch, err := repo.CurrentBranchName()
+	if err != nil {
+		return false, err
+	}
+
+	currentTree := root
+	for {
+		if currentTree.Branch.Name == currentBranch {
+			break
+		}
+		if len(currentTree.Next) > 1 {
```

Comment:
when would this be a case? I'm assuming we are not going to fail on the condition when root (main) has multiple branches as children which is pretty much always the case. Is current tree representation of root to current branch? If so, when can Next be > 1?

Reply:
The trunk branch (`main`/`master`/etc) is never considered a root. The root refers to the first off-trunk branch. So if we have:
```
main
  bookstore-backend
    bookstore-frontend
  petshop-backend
    petshop-frontend
```
the stack roots are `bookstore-backend` and `petshop-backend`.

Reply:
technically, even in that case, we should not have raise an error like this. So if you are doing sync in such as  setup:
```
main
   first
     first-x
     first-y
        first-y-1
        first-y-2 *
  
```
and you can av stack sync from `first-y-2`, we should sync `first` to `first-y` and `first-y` to `first-y-2`.